ghcide/src/Development/IDE/LSP/Notifications.hs

```diff
@@ -38,16 +38,21 @@ import           Development.IDE.Core.RuleTypes        (GetClientSettings (..))
 import           Development.IDE.Types.Shake           (toKey)
 import           Ide.Plugin.Config                     (CheckParents (CheckOnClose))
 import           Ide.Types
+import           System.FilePath                       (takeExtension)
 
-whenUriFile :: Uri -> (NormalizedFilePath -> IO ()) -> IO ()
-whenUriFile uri act = whenJust (LSP.uriToFilePath uri) $ act . toNormalizedFilePath'
+whenUriHaskellFile :: MonadIO m => Uri -> (NormalizedFilePath -> m ()) -> m ()
+whenUriHaskellFile uri act = whenJust maybeHaskellFile $ act . toNormalizedFilePath'
+  where
+    maybeHaskellFile = do
+      fp <- LSP.uriToFilePath uri
+      if takeExtension fp `elem` [".hs", ".lhs"] then Just fp else Nothing
```

Comment:
This does not seem like a very good solution since you have remember to use this in the "default" case.

Reply:
One day we will want to track changes and other notifications in `.cabal` files, and this code will need changing. That day might actually be today if we decide to generalise the plugin model to work with arbitrary extensions, as suggested by @fendor. In that case, a more appropriate solution would be to track notifications for any URIs involving file extensions which have associated plugins (specifically, plugins with non-empty rules).